### PR TITLE
support for nested properties in PartTree methods

### LINKED ISSUE #2305: Datastore - add nested properties support for PartTree methods

### PR DESCRIPTION
**Pull request by dmitry-s**
fixes #2305

### CODE REVIEW

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java [hunk 1]
```diff
@@ -319,11 +320,20 @@ else if (pageable instanceof DatastorePageable) {
...
 		Filter[] filters = this.filterParts.stream().map((part) -> {
-			DatastorePersistentProperty persistentProperty = (DatastorePersistentProperty) this.datastorePersistentEntity
-					.getPersistentProperty(part.getProperty().getSegment());
+			Iterable<PropertyPath> iterable = () -> part.getProperty().iterator();
+			List<DatastorePersistentProperty> properties =
+					StreamSupport.stream(iterable.spliterator(), false)
+					.map(propertyPath -> {
+						DatastorePersistentEntity<?> persistentEntity =
+								this.datastoreMappingContext.getPersistentEntity(propertyPath.getOwningType());
+						return persistentEntity.getPersistentProperty(propertyPath.getSegment());
```
Comment (elefeint) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java:329]: Will this be easier to read as a helper method? 
`.map(this::findProperty)`

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java [hunk 2]
```diff
@@ -335,9 +345,9 @@ private void applySelectWithFilter(Object[] parameters, Builder builder) {
 						"Too few parameters are provided for query method: " + getQueryMethod().getName());
 			}
 
-			Value convertedValue = convertParam(persistentProperty, it.next());
+			Value convertedValue = convertParam(properties.get(properties.size() - 1), it.next());
```
Comment (elefeint) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java:348]: Is there possibility of property name clashes if parent class and child class both have a property named, say, `name`?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java [hunk 3]
```diff
@@ -319,11 +320,20 @@ else if (pageable instanceof DatastorePageable) {
...
+			Iterable<PropertyPath> iterable = () -> part.getProperty().iterator();
+			List<DatastorePersistentProperty> properties =
+					StreamSupport.stream(iterable.spliterator(), false)
+					.map(propertyPath -> {
+						DatastorePersistentEntity<?> persistentEntity =
+								this.datastoreMappingContext.getPersistentEntity(propertyPath.getOwningType());
+						return persistentEntity.getPersistentProperty(propertyPath.getSegment());
+					}).collect(Collectors.toList());
+
+			String fieldName = properties.stream().map(DatastorePersistentProperty::getFieldName)
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java:332]: In Firestore you call the method `getPropertyName` but here it's `getFieldName`. Is it too late to make them consistent?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java [hunk 4]
```diff
@@ -319,11 +320,20 @@ else if (pageable instanceof DatastorePageable) {
 	private void applySelectWithFilter(Object[] parameters, Builder builder) {
 		Iterator it = Arrays.asList(parameters).iterator();
 		Filter[] filters = this.filterParts.stream().map((part) -> {
-			DatastorePersistentProperty persistentProperty = (DatastorePersistentProperty) this.datastorePersistentEntity
-					.getPersistentProperty(part.getProperty().getSegment());
+			Iterable<PropertyPath> iterable = () -> part.getProperty().iterator();
+			List<DatastorePersistentProperty> properties =
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java:324]: Put this block into a separate helper private method `buildName` like you did for Firestore?

**Pull request merged**